src/qibo/models/grover.py

```diff
@@ -0,0 +1,194 @@
+import numpy as np
+from qibo import gates
+from qibo.config import log, raise_error
+from qibo.models.circuit import Circuit
+
+
+class Grover(object):
+    """Model that performs Grover's algorithm.
+
+    # TODO: Add a few more details on Grover's algorithm and/or reference
+    for example `arXiv:quant-ph/9605043 <https://arxiv.org/abs/quant-ph/9605043>`_
+
+    Args:
+        oracle (:class:`qibo.core.circuit.Circuit`): quantum circuit that flips
+            the sign using a Grover ancilla initialized with -X-H-
+            and expected to have the total size of the circuit.
+        superposition_circuit (:class:`qibo.core.circuit.Circuit`): quantum circuit that
+            takes an initial state to a superposition. Expected to use the first
+            set of qubits to store the relevant superposition.
+        initial_state_circuit (:class:`qibo.core.circuit.Circuit`): quantum circuit
+            that initializes the state. If empty defaults to |000..00>
+        superposition_qubits (int): number of qubits that store the relevant superposition.
+            Leave empty if superposition does not use ancillas.
+        superposition_size (int): how many states are in a superposition.
+            Leave empty if its an equal superposition of quantum states.
+        number_solutions (int): number of expected solutions. Needed for normal Grover.
+            Leave empty for iterative version.
+        check (function): function that returns True if the solution has been
+            found. Required of iterative approach.
+            First argument should be the bitstring to check.
+        check_args (tuple): arguments needed for the check function.
+            The found bitstring not included.
+        iterative (bool): force the use of the iterative Grover
+    """
+
+    def __init__(self, oracle, superposition_circuit=None, initial_state_circuit=None,
+                 superposition_qubits=None, superposition_size=None, number_solutions=None,
+                 check=None, check_args=(), iterative=False):
+
+        self.oracle = oracle
+        self.initial_state_circuit = initial_state_circuit
+
+        if superposition_circuit:
+            self.superposition = superposition_circuit
+        else:
+            if not superposition_qubits:
+                raise_error(ValueError, "Cannot create Grover model if the "
+                                        "superposition circuit or number of "
+                                        "qubits is not specified.")
+            self.superposition = Circuit(superposition_qubits)
+            self.superposition.add([gates.H(i) for i in range(superposition_qubits)])
+
+        if superposition_qubits:
+            self.sup_qubits = superposition_qubits
+        else:
+            self.sup_qubits = self.superposition.nqubits
+
+        if superposition_size:
+            self.sup_size = superposition_size
+        else:
+            self.sup_size = int(2 ** self.superposition.nqubits)
```

Comment:
Is this `if` check required? Are there cases where the user has to pass a superposition size that is different than `2 ** nqubits`?

Reply:
Yes, although not commonly, one can search over the space of quantum states with, for example, k qubits out of n in state |1>. Although this could be worded differently, as its role is to get the number of Grover iterations to be performed, which technically its 1/amplitude of the solution state. This amplitude si most of the time 1/sqrt(superposition_size), and commonly 1/sqrt(2**superposition_qubits).